Signal/src/ViewControllers/HomeView/HomeViewCell.m

```diff
@@ -0,0 +1,424 @@
+}
+
++ (CGFloat)rowHeight
+{
+    return kHomeViewCellHeight;
+}
+
+- (void)initializeLayout
+{
+    self.selectionStyle = UITableViewCellSelectionStyleDefault;
+}
+
+- (nullable NSString *)reuseIdentifier
+{
+    return NSStringFromClass(self.class);
+}
+
+- (void)configureWithThread:(TSThread *)thread
+              contactsManager:(OWSContactsManager *)contactsManager
+        blockedPhoneNumberSet:(NSSet<NSString *> *)blockedPhoneNumberSet
+    shouldHaveBottomSeparator:(BOOL)shouldHaveBottomSeparator
+{
+    OWSAssertIsOnMainThread();
+    OWSAssert(thread);
+    OWSAssert(contactsManager);
+    OWSAssert(blockedPhoneNumberSet);
+
+    // TODO: Honor shouldHaveBottomSeparator.
+
+    self.thread = thread;
+    self.contactsManager = contactsManager;
+
+    BOOL hasUnreadMessages = thread.hasUnreadMessages;
+
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(otherUsersProfileDidChange:)
+                                                 name:kNSNotificationName_OtherUsersProfileDidChange
+                                               object:nil];
+    [self updateNameLabel];
+    [self updateAvatarView];
+
+    // We update the fonts every time this cell is configured to ensure that
+    // changes to the dynamic type settings are reflected.
+    self.snippetLabel.font = [self snippetFont];
+    self.snippetLabel.attributedText =
+        [self attributedSnippetForThread:thread blockedPhoneNumberSet:blockedPhoneNumberSet];
+    self.dateTimeLabel.attributedText = [self attributedStringForDate:thread.lastMessageDate];
+
+    self.separatorInset
+        = UIEdgeInsetsMake(0, kHomeViewAvatarSize + kHomeViewCellHMargin + kHomeViewAvatarHSpacing, 0, 0);
+
+    self.dateTimeLabel.textColor = hasUnreadMessages ? [UIColor ows_materialBlueColor] : [UIColor ows_darkGrayColor];
+
+    NSUInteger unreadCount = [[OWSMessageUtils sharedManager] unreadMessagesInThread:thread];
+    if (unreadCount > 0) {
+
+        self.unreadBadge.hidden = NO;
+        self.unreadLabel.font = [UIFont ows_dynamicTypeCaption1Font];
+        self.unreadLabel.text = [OWSFormat formatInt:MIN(99, (int)unreadCount)];
+
+        // TODO: Will this localize?  It assumes that the worst case
+        // unread count (99) will fit horizontally into some multiple
+        // N of the font's line height.
+        const int unreadBadgeSize = (int)ceil(self.unreadLabel.font.lineHeight * 1.5f);
+        self.unreadBadge.layer.cornerRadius = unreadBadgeSize / 2;
+
+        [self.viewConstraints addObjectsFromArray:@[
+                                                    [self.unreadBadge autoSetDimension:ALDimensionWidth toSize:unreadBadgeSize],
+                                                    [self.unreadBadge autoSetDimension:ALDimensionHeight toSize:unreadBadgeSize],
+                                                    [self.unreadBadge autoPinLeadingToTrailingEdgeOfView:self.payloadView offset:4.f],
```

Comment:
ha, I feel like I'm in the cult of the stack view all of a sudden... but... I've found this pattern is often nicely expressed with somethign like:

```
var views = [avatarView, payloadView]
if (unreadCount > 0) {
   views.append(unreadBadge)
}
contentStackView = [UIStackView withViews:views]
```

TIOLI obiviously

Reply:
I'm reserving judgment until I've used `UIStackView` more.  Layout is hard.  So far, it seems to work well though.